### PR TITLE
receive: Replication

### PR DESCRIPTION
**Pull request by squat**
This commit adds a new replication feature to the Thanos receiver.
By default, replication is turned off, however, when replication is
enabled (replication factor >=2), the target node for a time series will
replicate the time series to the other nodes concurrently and
synchronously. If the replication to >= (rf+1)/2 nodes fails, the
original write request is failed.

## Changes

* add optional replication to the Thanos receive component
* add e2e tests to exercise replication

## Verification

* e2e tests
* manual testing to ensure replicated metrics appear in querier
* checking new metrics to ensure replication errors don't increase in regular operation

cc @brancz @bwplotka @metalmatze

### CODE REVIEW

**Comment by brancz** [comment]
As replication wasn't specified in the original design, could you move the paragraph from open questions to the "Rollout/scaling/failure of receiver nodes" section and add appropriate detail?

File: pkg/receive/handler.go [hunk 1]
```diff
@@ -245,17 +313,37 @@ func (h *Handler) forward(ctx context.Context, tenant string, wreq *prompb.Write
...
 
+	var errs error
+	// Collect all the errors into a single error.
+	for _, err := range h.parallelizeRequests(ctx, tenant, replicas, wreqs, forwardRequestsTotal, forwardErrorsTotal) {
+		if err != nil {
+			if errs == nil {
+				errs = err
+				continue
+			}
+			errs = errors.Wrap(errs, err.Error())
```
Comment (brancz) [inline pkg/receive/handler.go]: MultiErrror makes more sense here no? In fact `.parallelizeRequsets` could immediately return a MultiError

File: pkg/receive/handler.go [hunk 2]
```diff
@@ -49,16 +50,42 @@ var (
 		},
 		[]string{"handler"},
 	)
+	replicationRequestsTotal = prometheus.NewCounter(
```
Comment (brancz) [inline pkg/receive/handler.go]: let's move not only the registration, but also the initialization of these to the `NewHandler` function

File: pkg/receive/handler.go (hunk 1 above)
Comment (squat) [reply pkg/receive/handler.go]: Sounds good! I didn’t realize we had a util for that :)

File: pkg/receive/handler.go (hunk 2 above)
Comment (metalmatze) [reply pkg/receive/handler.go]: Then we should also pass the `prometheus.Registry`?!

File: pkg/receive/handler.go [hunk 3]
```diff
@@ -288,13 +373,30 @@ func (h *Handler) forward(ctx context.Context, tenant string, wreq *prompb.Write
 				return
 			}
 			req.Header.Add(h.options.TenantHeader, tenant)
+			req.Header.Add(h.options.ReplicaHeader, strconv.FormatUint(replicas[endpoint], 10))
+
+			// Increment the counters as necessary now that
+			// the requests will go out.
+			defer func() {
+				requestCounter.Inc()
```
Comment (metalmatze) [inline pkg/receive/handler.go]: This will also be increased, even when there is an error. Is that intended?

File: pkg/receive/handler.go (hunk 2 above)
Comment (brancz) [reply pkg/receive/handler.go]: Precisely.

File: pkg/receive/handler.go (hunk 3 above)
Comment (squat) [reply pkg/receive/handler.go]: yes, this is tracking the _total_, number of requests. failed or successful. Using the error counter we can know the percentage of errors.

Comment (brancz) [reply pkg/receive/handler.go]: it's probably worth unifying in one metric with different labels for success/failure

Comment (metalmatze) [reply pkg/receive/handler.go]: Yep, I'd prefer having a unified counter. :+1:

Comment (squat) [reply pkg/receive/handler.go]: so don’t track total; instead use one counter vec with a result label for success or failure. When we want to know the percentage of failed request we do errors/(errors+successes)?

File: pkg/receive/handler.go (hunk 2 above)
Comment (squat) [reply pkg/receive/handler.go]: The registry is already passed into NewHandler via the options struct :+1:

File: pkg/receive/handler.go (hunk 3 above)
Comment (brancz) [reply pkg/receive/handler.go]: correct

File: docs/proposals/approved/201812_thanos-remote-receive.md [hunk 4]
```diff
@@ -157,11 +165,29 @@ To start, exact matches of tenant IDs will used to distribute requests to receiv
...
+
+### Replication
+
+The Thanos receiver supports replication of received time-series to other receivers in the same hashring. The replication factor is controlled by setting a flag on the receivers and indicates the maximum number of copies of any time-series that should be stored in the hashring. If any time-series in a write request received by a Thanos receiver is not successfully written to at least `(REPLICATION_FATOR + 1)/2` nodes, the receiver responds with an error. For example, to attempt to store 3 copies of every time-series and ensure that every time-series is successfully written to at least 2 Thanos receivers in the target hashring, all receivers should be configured with the following flag:
```
Comment (bwplotka) [inline docs/proposals/approved/201812_thanos-remote-receive.md]:
```suggestion
The Thanos receiver supports replication of received time-series to other receivers in the same hashring. The replication factor is controlled by setting a flag on the receivers and indicates the maximum number of copies of any time-series that should be stored in the hashring. If any time-series in a write request received by a Thanos receiver is not successfully written to at least `(REPLICATION_FACTOR + 1)/2` nodes, the receiver responds with an error. For example, to attempt to store 3 copies of every time-series and ensure that every time-series is successfully written to at least 2 Thanos receivers in the target hashring, all receivers should be configured with the following flag:
```

**Pull request merged**